Add shutdown handler for new launcher

rvl opened the pull request:
Relates to #1314.

# Overview

This adds a simpler way of ensuring clean shutdown of the wallet on windows (and linux), which doesn't require DaedalusIPC. The mechanism will be used by cardano-launcher.

- Adds the shutdown handler thread.
- Needed to rearrange startup functions around a little bit.
- Unit tests and an integration test.

# Comments

[Hydra jobset](https://hydra.iohk.io/jobset/Cardano/cardano-wallet-pr-1315)

rvl committed:
Add new simpler shutdown handler

rvl committed:
nix: add cardano-wallet-byron attribute for cardano-launcher shell

rvl commented on default.nix:
```diff
@@ -39,6 +39,8 @@ let
       haskellBuildUtils = iohkLib.haskellBuildUtils.package;
     };
 
+    inherit (haskellPackages.cardano-wallet-byron.components.exes) cardano-wallet-byron;
```
This is used by shell.nix in cardano-launcher.

paweljakubas commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Server.hs:
```diff
@@ -237,6 +244,30 @@ spec = do
+
+                withBackendProcessHandle nullTracer cmd $ \(Just hStdin) _ -> do
+                    threadDelay oneSecond
+                    hClose hStdin
```
**[question]**

it all comes down to this when shutting down on cmd, right? No way that we have some kind of shutdown without closing stdin handle?

rvl replied:
Yes good point - I added a check to make sure that the process exits with status 0.

paweljakubas reviewed (approved):
I was looking only at Haskell code only trying to understand logic, mainly in Startup and Launcher modules. Looks fine. Very fond of different cases covered in Cardano.StartupSpec. Looks like you nail it. Of course the ultimate verification is upon operation which to be seen.
btw. I see some complaining of weeder : 
```
warning: unknown setting 'sandbox-fallback'
```

KtorZ committed:
fix logging spec & retry mock-Daedalus request on connection failure

We send the "ReplyPort" actually just before the API is ready. So, it
can happen that the mock client fail to connect if it tries to connect
too quickly. The mock tests are now retrying once after 500ms in case
of connection failure.

-- Normal

JS: message received { ReplyPort: 38203 }
JS: child_process disconnected.
JS: waiting 2 seconds before exiting...
JS: child_process exited with status 0 or signal null

-- When the JS client connects too quickly

JS: message received { ReplyPort: 34083 }
JS: Connection refused. Retrying once in a bit...
JS: child_process disconnected.
JS: waiting 2 seconds before exiting...
JS: child_process exited with status 0 or signal null
